### PR TITLE
fix: --http-retry-max is ignored on TLS connections

### LINKED ISSUE #29188: `--http-retry-max` is silently ignored on TLS (non-plaintext) connections
# `--http-retry-max` is silently ignored on TLS (non-plaintext) connections

## Checklist

- [x] I've searched in the docs and FAQ for my answer.
- [x] I've included steps to reproduce the bug.
- [x] I've pasted the output of `argocd version`.

## Describe the bug

The global CLI flag `--http-retry-max` (and the equivalent `ClientOptions.HttpRetryMax`
in `pkg/apiclient`) has no effect whenever the client talks to a TLS server
(i.e. any non-`--plaintext` connection, which is the normal case). The
retryable transport is constructed and then immediately overwritten by the TLS
setup block, so requests are sent through a plain, non-retrying transport. A
single transient failure (e.g. an HTTP `502` from a load balancer in front of
`argocd-server`) is therefore returned to the caller as `fatal` on the first
attempt, with no retry.

This affects `--grpc-web` invocations in particular, because in grpc-web mode
every RPC is sent through `c.httpClient` (see `pkg/apiclient/grpcproxy.go`,
`executeRequest` → `c.httpClient.Do(req)`), which is exactly the client whose
retry behavior is being discarded.

## Root cause

In `pkg/apiclient/apiclient.go`, `NewClient` builds the retryable client and
then unconditionally replaces its `Transport` for TLS:

```go
if opts.HttpRetryMax > 0 {
    retryClient := retryablehttp.NewClient()
    retryClient.RetryMax = opts.HttpRetryMax
    c.httpClient = retryClient.StandardClient()   // Transport = *retryablehttp.RoundTripper
} else {
    c.httpClient = &http.Client{}
}

if !c.PlainText {
    tlsConfig, err := c.tlsConfig()
    if err != nil {
        return nil, err
    }
    c.httpClient.Transport = &http.Transport{      // <-- overwrites the retry RoundTripper
        TLSClientConfig: tlsConfig,
    }
}
```

`retryablehttp.Client.StandardClient()` returns an `*http.Client` whose retry
logic lives entirely in its `Transport` (`&retryablehttp.RoundTripper{Client: c}`).
The `if !c.PlainText` block then assigns a bare `&http.Transport{...}` over that
field, so on any TLS connection the retry wrapper is thrown away. The retry only
survives for `--plaintext` connections, where the TLS block is skipped.

### How it got here

This looks like an unintended interaction between two independent PRs, not a
deliberate design choice:

- #6444 (`fix: re-use same http transport in grpc proxy to avoid goroutine
  leak`) hoisted the per-request client construction out of `grpcproxy.go` into
  `NewClient` as `c.httpClient = &http.Client{}` followed by the `if !c.PlainText`
  TLS block. Correct in isolation — it was the single place the grpc-web proxy
  client was built.
- #6632 (`feat: make http client retryable`, closes #6600) later wrapped only
  the first line in `if opts.HttpRetryMax > 0 { ... StandardClient() }`, leaving
  the existing TLS block below it untouched. Because `StandardClient()` carries
  its retry logic in `Transport`, the pre-existing TLS block now silently
  clobbers it.

The regression is invisible unless a test asserts that `--http-retry-max`
actually produces a retrying transport against a *TLS* server; the motivating
issue (#6600) was about connection failures on unreliable networks and there is
no test exercising the TLS path.

## To Reproduce

```go
c, _ := apiclient.NewClient(&apiclient.ClientOptions{
    ServerAddr:   "argocd.example.com:443", // TLS (not --plaintext)
    HttpRetryMax: 4,
})
// c.(*client).httpClient.Transport is *http.Transport, NOT *retryablehttp.RoundTripper
```

Or in practice, against an `argocd-server` fronted by a load balancer that can
return a transient `502`:

```
argocd --grpc-web --http-retry-max 4 --server argocd.example.com \
  app set my-app --helm-set image.tag=<sha>
# fatal on first 502; no retry is attempted
{"level":"fatal","msg":"rpc error: code = Unknown desc = POST https://argocd.example.com/application.ApplicationService/UpdateSpec failed with status code 502"}
```

## Expected behavior

`--http-retry-max N` should retry retryable HTTP failures (including 5xx such as
`502`) on TLS connections, not only on `--plaintext` connections.

## Version

Reproduced by code inspection against `master`; the offending block is unchanged
from v3.1.x through the current `master`. Example failing invocation observed
with:

```
argocd: v3.1.1
```

## Proposed fix

Set the TLS config on the retryable client's own transport *before* wrapping it,
and only fall back to overwriting `Transport` in the non-retry path. See
accompanying PR. This keeps the goroutine-leak fix from #6444 intact (single
shared client, TLS honored) while preserving the retry transport from #6632.

### PR DESCRIPTION
**Pull request by mooselydev**
## What does this PR do / why do we need it?

The global CLI flag `--http-retry-max` (and `ClientOptions.HttpRetryMax` in
`pkg/apiclient`) has no effect on any TLS (non-`--plaintext`) connection, which
is the normal case. `NewClient` builds a retryable HTTP client and then
immediately overwrites its `Transport` in the TLS-setup block, discarding the
retry behavior. A single transient failure (e.g. a `502` from a load balancer
in front of `argocd-server`) is therefore returned to the caller as `fatal` on
the first attempt, with no retry.

This is especially impactful for `--grpc-web` invocations, because in grpc-web
mode every RPC is sent through `c.httpClient` (`pkg/apiclient/grpcproxy.go`,
`executeRequest` -> `c.httpClient.Do(req)`) -- exactly the client whose retry
behavior is being thrown away.

### Root cause

In `pkg/apiclient/apiclient.go`, `NewClient` had:

```go
if opts.HttpRetryMax > 0 {
    retryClient := retryablehttp.NewClient()
    retryClient.RetryMax = opts.HttpRetryMax
    c.httpClient = retryClient.StandardClient()   // Transport = *retryablehttp.RoundTripper
} else {
    c.httpClient = &http.Client{}
}

if !c.PlainText {
    tlsConfig, err := c.tlsConfig()
    if err != nil {
        return nil, err
    }
    c.httpClient.Transport = &http.Transport{      // overwrites the retry RoundTripper
        TLSClientConfig: tlsConfig,
    }
}
```

`retryablehttp.Client.StandardClient()` returns an `*http.Client` whose retry
logic lives entirely in its `Transport` (`&retryablehttp.RoundTripper{Client: c}`).
The `if !c.PlainText` block then assigns a bare `&http.Transport{...}` over that
field, so on any TLS connection the retry wrapper is silently discarded. Retry
only survived for `--plaintext` connections, where the TLS block is skipped.

This looks like an unintended interaction between two earlier changes rather
than a deliberate design choice: #6444 hoisted the shared HTTP client + TLS
block into `NewClient`, and #6632 later wrapped only the client construction in
`if opts.HttpRetryMax > 0 { ... StandardClient() }`, leaving the TLS block below
it untouched.

### Fix

Build the TLS transport first, then apply it to the retryable client's *inner*
`HTTPClient.Transport` before wrapping with `StandardClient()`, so both TLS and
retry are preserved. The non-retry path is behaviorally unchanged.

## Tests

Added to `pkg/apiclient/apiclient_test.go`:

- `TestNewClient_HttpRetryMax_TLSTransport` -- asserts the retry `RoundTripper`
  survives on a TLS connection and still carries a `TLSClientConfig`, and that
  the plain transport is used when `HttpRetryMax` is zero.
- `TestExecuteRequest_RetriesOn502` -- proves a transient 502 is retried through
  `executeRequest` (plain HTTP, focused on the retry mechanics).
- `TestNewClient_RetriesOn502_OverTLS` -- end-to-end: wires the client through
  `NewClient`, talks to a real `httptest.NewTLSServer`, and verifies a transient
  502 is retried over TLS and ultimately succeeds. This test fails against the
  pre-fix code (`expected retryablehttp.RoundTripper, got *http.Transport`).

## Checklist

* [x] Either (a) I've created an enhancement proposal and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the Title of the PR guidelines.
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them. (N/A -- no user-facing surface change; existing flag now works.)
* [ ] Does this PR require documentation updates? (No -- documented flag now behaves as documented.)
* [x] I've updated documentation as required by this PR. (N/A)
* [x] I have signed off all my commits as required by DCO.
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green.
* [x] My new feature complies with the feature status guidelines. (Bug fix.)
* [ ] Optional. My organization is added to USERS.md.
* [ ] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into.

Fixes #29188

### CODE REVIEW

File: pkg/apiclient/apiclient_test.go [hunk 1]
```diff
@@ -281,3 +282,116 @@ func (c *closeTracker) Close() error {
...
+		ServerAddr: server.URL[7:], // Remove "http://"
+		PlainText:  true,
+		httpClient: retryClient.StandardClient(),
+	}
+
+	ctx := t.Context()
+	md := metadata.New(map[string]string{})
+	_, err := c.executeRequest(ctx, "/test.Service/Method", []byte("test"), md)
+	require.NoError(t, err)
+	assert.Equal(t, int32(3), attempts.Load(), "expected two 502 retries before success")
```
[the remark this answers is not in the document]
Comment (mooselydev) [reply pkg/apiclient/apiclient_test.go:360]: Good catch — fixed in f869e63. Both `TestExecuteRequest_RetriesOn502` and `TestNewClient_RetriesOn502_OverTLS` now capture the response and `defer resp.Body.Close()`, matching the caller-closes-on-success contract documented by `TestExecuteRequest_SuccessDoesNotCloseBodyPrematurely`.

File: pkg/apiclient/apiclient.go [hunk 2]
```diff
@@ -263,23 +263,35 @@ func NewClient(opts *ClientOptions) (Client, error) {
...
 		}
-		c.httpClient.Transport = &http.Transport{
+		transport = &http.Transport{
 			TLSClientConfig: tlsConfig,
 		}
 	}
+
+	if opts.HttpRetryMax > 0 {
+		retryClient := retryablehttp.NewClient()
+		retryClient.RetryMax = opts.HttpRetryMax
```
Comment (choejwoo) [inline pkg/apiclient/apiclient.go:287]: Since retryablehttp.NewClient() comes with a default stderr logger, doesn't this mean requests that ultimately succeed still emit logs? a `DEBUG` line per request even when nothing is retried?
And the tests seem to work around this by setting `Logger = nil`, should NewClient do the same? What do you think?

Comment (mooselydev) [reply pkg/apiclient/apiclient.go:287]: Great catch, and yes — exactly right. `retryablehttp.NewClient()` sets a default stderr logger that logs `performing request` on every attempt, so with `--http-retry-max` set, successful non-retried calls would emit a line per request. The plain `http.Client` path is silent, so that'd be an inconsistent, noisy regression.

Fixed in 3ce0171: `NewClient` now sets `retryClient.Logger = nil`. I also replaced the test-side `Logger = nil` workaround with an assertion that `NewClient` disables it, so the behavior is covered rather than papered over. Thanks for the review!

File: pkg/apiclient/apiclient.go [hunk 3]
```diff
@@ -263,23 +263,40 @@ func NewClient(opts *ClientOptions) (Client, error) {
...
+		// retryablehttp.NewClient() installs a default logger that writes a line
+		// to stderr for every request, even successful non-retried ones. The
+		// plain http.Client path below is silent, so disable the logger to keep
+		// behavior consistent and avoid per-request noise on the CLI.
+		retryClient.Logger = nil
```
Comment (ppapapetrou76) [inline pkg/apiclient/apiclient.go:292]: please consider adding also this 

```go
retryClient.CheckRetry = retryablehttp.ErrorPropagatedRetryPolicy
```

without it, a persistent `502` error now reports `giving up after 5 attempt(s)` with the status code stripped, where before your change TLS users got `failed with status code 502`

File: pkg/apiclient/apiclient_test.go [hunk 4]
```diff
@@ -281,3 +282,123 @@ func (c *closeTracker) Close() error {
...
+		c := ci.(*client)
+
+		tr, ok := c.httpClient.Transport.(*http.Transport)
+		require.True(t, ok, "expected plain *http.Transport, got %T", c.httpClient.Transport)
+		require.NotNil(t, tr.TLSClientConfig)
+	})
+}
+
+// TestExecuteRequest_RetriesOn502 proves that with HttpRetryMax set, a
+// transient 502 is retried rather than returned fatally on the first attempt.
```
Comment (ppapapetrou76) [inline pkg/apiclient/apiclient_test.go]:
```suggestion
// TestExecuteRequest_RetriesOn502 covers executeRequest's handling of a response
// delivered through a retrying transport. The regression test for the TLS bug is
// TestNewClient_RetriesOn502_OverTLS.
```

File: pkg/apiclient/apiclient.go [hunk 5]
```diff
@@ -263,23 +263,40 @@ func NewClient(opts *ClientOptions) (Client, error) {
...
 	if !c.PlainText {
 		tlsConfig, err := c.tlsConfig()
 		if err != nil {
 			return nil, err
 		}
-		c.httpClient.Transport = &http.Transport{
+		transport = &http.Transport{
 			TLSClientConfig: tlsConfig,
 		}
 	}
```
Comment (ppapapetrou76) [inline pkg/apiclient/apiclient.go]:
```suggestion
      // *http.Transport rather than http.RoundTripper so we can close idle
      // connections on it directly (see grpcproxy.go).
      var transport *http.Transport
      if !c.PlainText {
              tlsConfig, err := c.tlsConfig()
              if err != nil {
                      return nil, err
              }
              transport = &http.Transport{
                      TLSClientConfig: tlsConfig,
              }
      }
      c.tlsTransport = transport
```

**Review by ppapapetrou76 — COMMENTED** [review]
Please consider adding something like this in grpcprocy

```go
            if c.tlsTransport != nil {
                  c.tlsTransport.CloseIdleConnections()
             } else {
                  c.httpClient.CloseIdleConnections()
             }
```

**Review by ppapapetrou76 — COMMENTED** [review]
Also how about adding something in the upgrade docs


`--http-retry-max previously` had no effect on TLS connections, which is the default. Since 3.6, be aware that retries apply to `--grpc-web` requests generally, including non-idempotent ones — a transient 5xx returned after the server has already processed a request will cause it to be re-sent. Consider this before enabling the flag for mutating commands.

File: pkg/apiclient/apiclient.go (hunk 3 above)
Comment (mooselydev) [reply pkg/apiclient/apiclient.go:292]: Done in 2d870f4 — set `retryClient.CheckRetry = retryablehttp.ErrorPropagatedRetryPolicy`. Confirmed a persistent 502 now surfaces as `...giving up after N attempt(s): unexpected HTTP status 502 Bad Gateway` instead of the status being stripped, and added `TestNewClient_PersistentError_PreservesStatus` to lock that in. Good catch — our downstream tooling keys on the status code.

File: pkg/apiclient/apiclient_test.go (hunk 4 above)
Comment (mooselydev) [reply pkg/apiclient/apiclient_test.go]: Applied in 2d870f4.

File: pkg/apiclient/apiclient.go (hunk 5 above)
Comment (mooselydev) [reply pkg/apiclient/apiclient.go]: Adopted in 2d870f4 — added a typed `c.tlsTransport *http.Transport` field, set it in NewClient, and used it in grpcproxy. You're right that this matters beyond style: `retryablehttp.RoundTripper` doesn't implement `CloseIdleConnections`, so on the retry path the existing `c.httpClient.CloseIdleConnections()` was a silent no-op (regressing the #6444 leak fix). grpcproxy now closes idle conns on `c.tlsTransport` when present. Added `TestNewClient_RetryPath_ExposesTLSTransport` to cover it.

**Comment by mooselydev** [comment]
Thanks for the thorough review @ppapapetrou76 — all points addressed in 2d870f4:

- **`ErrorPropagatedRetryPolicy`**: exhausted retries now preserve the status code (e.g. 502); added a test.
- **`c.tlsTransport` + grpcproxy `CloseIdleConnections`**: done — closing on `httpClient` was a no-op through the retryablehttp RoundTripper, so idle-conn cleanup on the retry path was silently lost; now closes on the underlying transport. Added a test.
- **Upgrade docs**: added a `v3.5 -> 3.6` note covering the behavior change and, per your point, the non-idempotent `--grpc-web` retry caveat (retries apply to all POST-tunneled RPCs; safe for declarative ops like `app set`/`app sync`, riskier for mutating commands). The flag still defaults to 0 (opt-in).
- Also capped `RetryWaitMax` at 10s so retries can't stack latency ahead of command timeouts.

On the non-idempotent path specifically: since the flag is opt-in and this risk predates the fix (it existed on the plaintext+retry path), I've documented it rather than trying to build an idempotency allowlist — gRPC-web erases the read/write distinction at the HTTP layer, so safe-by-default retry would be a larger design. Happy to pursue that separately if you'd prefer. CI is green apart from a flaky `Check changes to generated code` toolchain-install step (fails before codegen runs) — a re-run should clear it.
